### PR TITLE
Pluggable name generator

### PR DESCRIPTION
**Pull request by dain**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -49,7 +49,7 @@
         <dependency>
             <groupId>com.google.inject</groupId>
             <artifactId>guice</artifactId>
-            <version>3.0</version>
+            <version>4.2.2</version>
```
Comment (martint) [inline pom.xml]: The commit title says "Guava"

File: src/main/java/org/weakref/jmx/MBeanExporter.java [hunk 2]
```diff
@@ -35,23 +35,62 @@
 import java.util.Map;
```
Comment (martint) [inline src/main/java/org/weakref/jmx/MBeanExporter.java]: Typo in commit message: nameing -> naming

File: src/main/java/org/weakref/jmx/guice/MBeanModule.java [hunk 3]
```diff
@@ -37,6 +39,9 @@ protected final void configure()
         bind(GuiceMBeanExporter.class).asEagerSingleton();
         bind(MBeanExporter.class).in(Scopes.SINGLETON);
 
+        newOptionalBinder(binder(), ObjectNameGenerator.class)
+                .setDefault().toProvider(ObjectNameGenerator::defaultObjectNameGenerator);
```
Comment (martint) [inline src/main/java/org/weakref/jmx/guice/MBeanModule.java]: `.toProvider(...)` should go on next line

File: src/main/java/org/weakref/jmx/ObjectNameBuilder.java [hunk 4]
```diff
@@ -14,25 +14,26 @@
...
-        checkArgument(!BAD_PACKAGENAME_PATTERN.matcher(packageName).find(), "packageName is invalid");
-        this.objectName = new StringBuilder(packageName);
+        requireNonNull(domain, "domain is null");
+        checkArgument(!BAD_PACKAGENAME_PATTERN.matcher(domain).find(), "domain is invalid");
+        this.objectName = new StringBuilder(domain);
     }
 
     public ObjectNameBuilder withProperty(String name, String value)
     {
+        checkArgument(!properties.contains(name), "Duplicate property name '" + name + "'");
```
Comment (electrum) [inline src/main/java/org/weakref/jmx/ObjectNameBuilder.java]: String formatting
```java
checkArgument(!properties.contains(name), "Duplicate property name: %s", name);

**Comment by dain** [comment]
@martint I updated the following to use `Optional`  in the commit `Add pluggable ObjectNameGenerator for builtin naming`:
```
src/main/java/org/weakref/jmx/MBeanExporter.java
src/main/java/org/weakref/jmx/guice/GuiceMBeanExporter.java
src/main/java/org/weakref/jmx/guice/MBeanModule.java

**Pull request merged**